src/Features/CSharp/Portable/Diagnostics/Analyzers/CSharpUseImplicitTypeDiagnosticAnalyzer.cs

```diff
@@ -151,6 +154,57 @@ protected override bool TryAnalyzeVariableDeclaration(TypeSyntax typeName, Seman
             return false;
         }
 
+        private bool TryAnalyzeDeclarationExpression(
+            DeclarationExpressionSyntax declarationExpression,
+            SemanticModel semanticModel,
+            OptionSet optionSet,
+            CancellationToken cancellationToken)
+        {
+            // It's not always safe to convert a decl expression like "Method(out int i)" to 
+            // "Method(out var i)".  Changing to 'var' may cause overload resolution errors.  
+            // Have to see if using 'var' means not resolving to the same type as before.  
+            // Note: this is fairly expensive, so we try to avoid this if we can by seeing if
+            // there are multiple candidates with the original call.  If not, then we don't
+            // have to do anything.
+            if (declarationExpression.Parent is ArgumentSyntax argument &&
+                argument.Parent is ArgumentListSyntax argumentList &&
+                argumentList.Parent is InvocationExpressionSyntax invocationExpression)
+            {
+                var memberGroup = semanticModel.GetMemberGroup(invocationExpression.Expression, cancellationToken);
+                if (memberGroup.Length == 1)
+                {
+                    // Only one member found.  Don't do the expensive check, this change 
+                    // should be ok.
+                    return true;
+                }
+            }
+
+            // Do the expensive check.  Note: we can't use the SpeculationAnalyzer (or any 
```

Comment:
😨 (No changes required at this time for this comment. We are working on the ability to performance-test analyzers and scale, and any problems in this code will be revealed without question.)

I'm really interested to see where this breaks down in practice. There are many different optimizations we can do to avoid the expensive check, but it'd be awesome to focus on ones that make the most difference.

Examples:

* Argument list length
* Presence of an `out` variable at a specific position

Reply:
> and any problems in this code will be revealed without question.)

It really comes down to usage patterns around out-variables.  i.e. do we expect them to be used commonly in things like constructors.  

Now, that said, i would prefer the underlying issue with speculative analysis is fixed.  That's much more critical *especially* as patterns (and nested patterns) are going to be incredibly common.  We cannot properly speculatively analyze here because the compiler gets very unhappy and things the locals from before/after are colliding.  I think a suitable fix would be for the compiler to be ok with the introduction (in speculative code) of variables with the exact same type and name in the same statement.  That could also be controlled by a flag if someone wanted the existing semantics.